### ISSUE TITLE
DiscreteEvents: solve_ivp should evaluate fun() at specific, user-specified timepoints

### ISSUE THREAD

**Issue by alexlenail** [scipy/scipy#14246]
I'd like to force solve_ivp to evaluate the ODE RHS `fun()` at specific timepoints, in order to implement the adjoint method (as in [this paper](https://arxiv.org/pdf/1806.07366.pdf)). 

- Registering an event at specific timepoints with `solve_ivp` returns the ODE solution at those timepoints using interpolation, and skips evaluating `fun()` at those points. 
- Similarly, supplying `t_eval` to `solve_ivp` at those points _does not actually evaluate fun() at those t_, but instead, gives you back a solution with the interpolated values at those points. 

I'd like to be able to specify a set of timepoints at which the solver _must_ evaluate `fun()`.

**Comment by alexlenail** [comment]
Note: `LSODA` has an `events` argument for this purpose underneath, but it isn't exposed by the scipy wrapper: 

https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/integrate/lsoda.pyf#L27

However, R's deSolve and others _do_ expose this option (see `events` in https://www.rdocumentation.org/packages/deSolve/versions/1.28/topics/lsoda)

**Comment by laurent90git** [comment]
This is not implemented so far, however could be easily added I think. In the meantime, you can simply perform multiple integrations, each time setting the former endpoint as starting time and the new endpoint at the next time point where you want the RHS to be evaluated.

**Comment by drhagen** [comment]
I don't think forcing `solve_ivp` to evaluate the RHS at particular time points is going to be enough to do what you want. I did not read that particular paper, but when I implemented an adjoint method, it involved _altering the state_ at those particular time points. This is a much different beast. The solver cannot simply continue after such a change (i.e. do not directly manipulate the state vector of an adaptive ODE solver or your will  get wrong answers). The solver must be stopped and restarted at those points.

That being said, the ability to apply discontinuities was one of my original design requirements. I used the step solvers to [do just that](https://github.com/drhagen/biolucia/blob/master/biolucia/ode.py#L350-L417) in my experimental systems biology package Biolucia. Now, this uses the [step solvers directly](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.LSODA.html#scipy.integrate.LSODA) and not `solve_ivp`. For implementing the adjoint, you could either use the step solvers or just use `solve_ivp` to solve each interval, and manually apply the discontinuity between each interval.

I am a big believer in using the step solvers to do complex, arbitrary things and leaving `solve_ivp` to handle the simple case. However, I get the feeling that other people disagree. So if this were to get baked into `solve_ivp`, it should probably take an argument `discontinuities: List[float]`, which stops and restarts the solver at those particular times. Then there is another argument `jump: (t: float, x: Array[float, [nx]]) -> Array[float, [nx]]`, which is a function that is called at each discontinuity. The function takes the current time and current state and returns a new state. I guess `solve_ivp` would pass `*args` to this function also.

**Comment by alexlenail** [comment]
@drhagen you're right about the adjoint method, I had been mis-reading the math. I've moved on to using diffeqpy for my particular project, but would still love to see this in scipy.

**Comment by MatthewFlamm** [comment]
I'm not sure whether this request is really about forcing integrations to hit specific points or about adjoint methods, but if the former then #23385 would close this issue.
